### ISSUE TITLE
about dtw.py

### ISSUE THREAD

**Issue by OvOYu** [Mouradost/STGM#10]
The purpose of dtw.py is to generate a contribution matrix, but how to execute the dtw.py file? Can you give an example to execute the dtw.py file through different data sets? Do you need other codes?

**Comment by Mouradost** [comment]
Everything is explained in [the README under run STGM or your custom version](https://github.com/Mouradost/STGM#-run-stgm-or-your-custom-version-), you provide `dtw.py` with the path to the signal data, and it will construct the similarity matrix and no you do not need other codes. You can also train without pre-generating the similarity matrix the `train.py` will auto-detect that the similarity matrix is not provided and will be generated on the fly while training.

**Comment by ShihPing0803** [comment]
Hello @Mouradost, I use the command `python dtw.py "C:\Users\yzu_im_4070\Desktop\STGM\datasets\pemsbay\pems-bay.h5"`. When I run `run.py` to train it, it just shows the following script, I think my dtw doesn't work. 
Is DTW used for the circled part in the image below?
<img width="523" alt="螢幕擷取畫面 2023-08-07 192609" src="https://github.com/Mouradost/STGM/assets/119124535/6ef39ffe-8184-4e1f-98a3-8d14c7ffad9b">
<img width="543" alt="螢幕擷取畫面 2023-08-07 192741" src="https://github.com/Mouradost/STGM/assets/119124535/7f226be8-de8c-4fec-b6c7-a6768f95411e">
<img width="368" alt="螢幕擷取畫面 2023-08-07 192914" src="https://github.com/Mouradost/STGM/assets/119124535/3deeb919-4323-4f37-90e6-087958506b42">

---script---
C:\Users\yzu_im_4070\Desktop\STGM\src>python run.py trainer.epochs=100 trainer.batch_size=128 dataset=pemsbay
[2023-08-07 19:13:09,936][Run][INFO] - Logs will be saved at C:\Users\yzu_im_4070\Desktop\STGM\src\logs\PEMS-BAY\STGM\2023-08-07\19-13-09
[2023-08-07 19:13:09,956][Run][INFO] - Using cuda device -> NVIDIA GeForce RTX 4070 Ti
[2023-08-07 19:13:10,004][numexpr.utils][INFO] - Note: NumExpr detected 32 cores but "NUMEXPR_MAX_THREADS" not set, so enforcing safe limit of 8.
[2023-08-07 19:13:10,004][numexpr.utils][INFO] - NumExpr defaulting to 8 threads.
[2023-08-07 19:13:10,496][Dataset][INFO] - Dataset PEMSBAY (train) length: 36457
[2023-08-07 19:13:10,497][Dataset][WARNING] - No pre-calculated contribution matrix was provided ! You will experience Heavy CPU load and Slow data loading...
[2023-08-07 19:13:10,784][Dataset][INFO] - Dataset PEMSBAY (val) length: 5187
[2023-08-07 19:13:10,785][Dataset][WARNING] - No pre-calculated contribution matrix was provided ! You will experience Heavy CPU load and Slow data loading...
[2023-08-07 19:13:10,785][Run][INFO] - Dataset PEMSBAY mean: 62.619574 std: 9.594371
Epoch 0/100 train:   0%|                                                                       | 0/100 [00:00<?, ?it/s]

**Comment by Mouradost** [comment]
@loading2022 Yes it takes the `metr-la.h5` and performs the DTW algorithm to create the similarity matrix.
I don't see any problem with your run, reading your log it seems that the train successfully started. If you are worried about the `[2023-08-07 19:13:10,785][Dataset][WARNING] - No pre-calculated contribution matrix was provided ! You will experience Heavy CPU load and Slow data loading...` this is normal, it's because you didn't calculate the similarity matrix beforehand, so the training will calculate the similarity while loading the data in the VRAM.

**Comment by ShihPing0803** [comment]
> @loading2022 Yes it takes the `metr-la.h5` and performs the DTW algorithm to create the similarity matrix.
> I don't see any problem with your run, reading your log it seems that the train successfully started. If you are worried about the `[2023-08-07 19:13:10,785][Dataset][WARNING] - No pre-calculated contribution matrix was provided ! You will experience Heavy CPU load and Slow data loading...` this is normal, it's because you didn't calculate the similarity matrix beforehand, so the training will calculate the similarity while loading the data in the VRAM. 

I run `python dtw.py` first, and then run `python run py`. Then, it appears that `No pre-calculated contribution matrix was provided!`

**Comment by Mouradost** [comment]
Because you didn't provide it to the dataset for that you need to rewrite the class under `src/dataset/[YOUR-DATASET-NAME.py]` and load it to the internal variable `self.sim`

**Comment by ShihPing0803** [comment]
I saw in `dtw.py` line 144 written as "The saved DTW numpy file name. [str] (default: dtw)", will there be a file after running it?

**Comment by OvOYu** [comment]
I run the dtw.py file directly and follow the path of the h5 file. I can see that the dtw.npy file can be generated. I didn’t rewrite it. Is this correct?

**Comment by OvOYu** [comment]
Is it right to do this? First run the dtw file that generates 3 data, and then run run.py. I see that the dtw file seems to be uniformly named dtw.npy. How should I rename them?

**Comment by Mouradost** [comment]
The `dtw.py` generate a numpy `.npy` file which represents the contribution of each node over the other during each period (the period can be selected to match each timesteps (12 steps as an exemple) or any arbitrary temporal size of window even the whole dataset) this file can than be used during training using the internal variable of the dataset class `self.sim`. This step is only a tool meant  to speed up training considerably. However, if not performed and the `self.sim` not provided that doesn't affect anything except the time it takes to finish training, since the contribution matrix is calculated on the fly anyway if not provided.

**Comment by ShihPing0803** [comment]
Thanks for your response!

**Comment by ShihPing0803** [comment]
For example, I need to add `self.sim` in [datasets].py
```
# ML
import pandas as pd

# Own
from datasets.base import BaseDataset


class Dataset(BaseDataset):
    def load_data(self):
        self._name = "PEMSBAY"
        self.adj = pd.read_csv(
            self.data_folder_path / "PEMSBAY" / "W_pemsbay.csv"
        ).values
        data = pd.read_hdf(self.data_folder_path / "PEMSBAY" / "pems-bay.h5")
        self.timestamps = data.index
        self.timestamps.freq = self.timestamps.inferred_freq
        self.time_index = self.time_to_idx(self.timestamps, freq="5min")
        self.data = data.values
        self.data_mean, self.data_std = self.data.mean(), self.data.std()
        self.data_min, self.data_max = self.data.min(), self.data.max()
        self.sim=np.load('dtw.npy')
```
Is it correct for adding similarity matrix?

**Comment by Mouradost** [comment]
Yes, don't forget to import numpy.

**Comment by OvOYu** [comment]
thank you for your reply

**Comment by ShihPing0803** [comment]
Hello, I have another question to ask. If I don't run `dtw.py` and directly run `run.py`, will it generate `dtw.npy` automatically? Also, which segment of code in `run.py` is related to generating the DTW?

**Comment by Mouradost** [comment]
Yes it will, but will not save the DTW matrix into a file such as `dtw.npy` and the related code for generating the DTW on the fly can be found [here](https://github.com/Mouradost/STGM/blob/main/src/datasets/base.py#L154)

**Comment by ShihPing0803** [comment]
I got it. Thank you!

**Comment by OvOYu** [comment]
![image](https://github.com/Mouradost/STGM/assets/107761563/d6042e25-7113-4c5f-bf15-21c5a0f70573)
With your help, I successfully generated those two npy files, but regarding PEMSD7M, I configured a csv file and prompted an error. May I ask which file you should choose? I see that the code requirement must be an h5 file

**Comment by Mouradost** [comment]
@OvOYu Indeed the DTW code shared in this repos only supports `h5` files. I will provide an update in the future to accept both `csv` and `h5` files but not in the near future given that I'm a little bit busy with my current work. Meanwhile, you are free to contribute to this repos if you want to implement this feature yourself I will accept PRs 😉.

**Comment by OvOYu** [comment]
Thank you very much for your reply and wish you success in your work

**Comment by Mouradost** [comment]
Thanks hope you will succeed as well, and good luck with your research 😉😊

**Comment by OvOYu** [comment]
D:\python\anaconda\Anaconda\envs\310p\python.exe C:/Users/admin/Documents/STGM-main/src/run.py trainer.epochs=100 trainer.batch_size=128 dataset=metrla device=cuda:0 model.nb_blocks=1 model=stgm log.logger=wandb
wandb: Currently logged in as: ovoyu (ovoshuai). Use `wandb login --relogin` to force relogin
wandb: Tracking run with wandb version 0.15.8
wandb: Run data is saved locally in C:\Users\admin\Documents\STGM-main\src\logs\METR-LA\STGM\2023-08-09\09-32-05\wandb\run-20230809_093207-ua7fb85o
wandb: Run `wandb offline` to turn off syncing.
wandb: Syncing run STGM
wandb:  View project at https://wandb.ai/ovoshuai/TrafficForcastingSTGM
wandb:  View run at https://wandb.ai/ovoshuai/TrafficForcastingSTGM/runs/ua7fb85o
[2023-08-09 09:32:19,239][Run][INFO] - Logs will be saved at C:\Users\admin\Documents\STGM-main\src\logs\METR-LA\STGM\2023-08-09\09-32-05
[2023-08-09 09:32:19,278][Run][INFO] - Using cuda:0 device -> NVIDIA GeForce RTX 3090
[2023-08-09 09:32:19,470][numexpr.utils][INFO] - Note: NumExpr detected 32 cores but "NUMEXPR_MAX_THREADS" not set, so enforcing safe limit of 8.
[2023-08-09 09:32:19,470][numexpr.utils][INFO] - NumExpr defaulting to 8 threads.
[2023-08-09 09:32:20,207][Dataset][INFO] - Dataset METRLA (train) length: 23966
[2023-08-09 09:32:20,489][Dataset][INFO] - Dataset METRLA (val) length: 3403
[2023-08-09 09:32:20,490][Run][INFO] - Dataset METRLA mean: 53.719021 std: 20.261431
Epoch 100/100 train: 100%|██████████| 100/100 [55:32<00:00, 33.32s/it, e_loss=0.828, e_loss_val=0.604, loss=25.4, loss_val=21.7]
D:\python\anaconda\Anaconda\envs\310p\lib\site-packages\plotly\matplotlylib\renderer.py:611: UserWarning:

I found a path object that I don't think is part of a bar chart. Ignoring.

[2023-08-09 10:27:54,029][STGM][INFO] - Model succesfully saved
[2023-08-09 10:27:54,248][Dataset][INFO] - Dataset METRLA (test) length: 6831
Generating predictions: 100%|██████████| 214/214 [00:04<00:00, 43.38it/s]
[2023-08-09 10:27:59,967][root][INFO] - 

================================================================================
                                Mean Error Real                                 
--------------------------------------------------------------------------------
        MSE                 RMSE                MAE                 MAPE        
--------------------------------------------------------------------------------
       248.22              15.75               10.53               32.09%       
================================================================================
================================================================================
                               By Step Error Real                               
--------------------------------------------------------------------------------
     Steps            MSE             RMSE            MAE             MAPE      
--------------------------------------------------------------------------------
       1             264.76          16.27            9.37           31.89%     
       2             240.49          15.51           10.36           31.80%     
       3             270.05          16.43           12.88           33.85%     
       4             241.18          15.53            9.64           31.25%     
       5             250.06          15.81           11.65           32.78%     
       6             254.33          15.95            9.37           31.65%     
       7             245.35          15.66           10.77           32.29%     
       8             242.22          15.56           10.40           31.92%     
       9             238.75          15.45           10.26           31.66%     
       10            246.58          15.70           11.35           32.50%     
       11            241.50          15.54           10.38           31.88%     
       12            243.39          15.60            9.89           31.60%     
================================================================================

wandb: Waiting for W&B process to finish... (success).
wandb:  View run STGM at: https://wandb.ai/ovoshuai/TrafficForcastingSTGM/runs/ua7fb85o
wandb: Synced 6 W&B file(s), 1 media file(s), 0 artifact file(s) and 0 other file(s)
wandb: Find logs at: .\wandb\run-20230809_093207-ua7fb85o\logs

Process finished with exit code 0


Hello, this is my training log on metrla. I observed that the error is much worse than what you reported in the experiment. Please tell me how to solve this situation, thank you

**Comment by Mouradost** [comment]
hi, read this part of the [README](https://github.com/Mouradost/STGM#-disclaimer-).
here are some solutions that might help:
- Try to follow the hyper-parameters in the papers.
- The estimator is sometimes unstable which might influence the training of the two models, first train STGM without the estimator if the error is similar to the papers, include the estimator and freeze STGM.
- Train both models separately, then together for few extra steps.
- Training for only 5 or 10 epoch and observe if the error is reducing correctly (rapide reduction in the few first steps) and continue only if it's the case.

If nothing in the above have worked let me know, it might be due to a bug in the custom metrics or the implementation of the models.

**Comment by OvOYu** [comment]
thanks， good luck！

**Comment by Mouradost** [comment]
@OvOYu There was a bug in the prediction part. Now it's fixed. Update your code with the newly pushed code.
If you still have performance issues, let me know.
